dpnp/backend/kernels/dpnp_krnl_random.cpp

```diff
@@ -533,6 +533,292 @@ void dpnp_rng_negative_binomial_c(void* result, const double a, const double p,
+                dpnp_memory_free_c(idx);
+            }
+            else
+            {
+                for (i = 0; i < size; i++)
+                {
+                    mkl_rng::gamma<_DataType> gamma_distribution(shape + pvec[i], d_zero, d_two);
+                    event_out = mkl_rng::generate(gamma_distribution, DPNP_RNG_ENGINE, 1, result1 + 1);
+                    event_out.wait();
+                }
+            }
+            dpnp_memory_free_c(pvec);
+        }
+        else
+        {
+            /* noncentral_chisquare(1, nonc) ~ (Z + sqrt(nonc))**2 for df == 1 */
+            loc = sqrt(nonc);
+            mkl_rng::gaussian<_DataType> gaussian_distribution(loc, d_one);
+            event_out = mkl_rng::generate(gaussian_distribution, DPNP_RNG_ENGINE, size, result1);
+            event_out.wait();
+            event_out = mkl_vm::sqr(DPNP_QUEUE, size, result1, result1, no_deps, mkl_vm::mode::ha);
+            event_out.wait();
+        }
+    }
+    else
+    {
+        double shape, loc;
+        int errcode;
+        size_t i;
+
+        if (df > 1)
+        {
+            double* nvec = nullptr;
+
+            shape = 0.5 * (df - 1.0);
+            /* res has chi^2 with (df - 1) */
+            errcode =
+                vdRngGamma(VSL_RNG_METHOD_GAMMA_GNORM_ACCURATE, get_rng_stream(), size, result1, shape, d_zero, d_two);
+            // TODO
+            // refactor this check, smth like: void status_assert(errcode, VSL_STATUS_OK, func_name)
+            // or just redesign and return int status:
+            // void dpnp_rng_*_c(..., int& status);
+            // or
+            // int dpnp_rng_*_c(...);
+            if (errcode != VSL_STATUS_OK)
+            {
+                throw std::runtime_error("DPNP RNG Error: dpnp_rng_noncentral_chisquare_c() failed.");
+            }
+
+            nvec = (double*)mkl_malloc(size * sizeof(double), 64);
+            if (nvec == nullptr)
+            {
+                throw std::runtime_error("DPNP RNG Error: dpnp_rng_noncentral_chisquare_c() failed.");
+            }
+            loc = sqrt(nonc);
+            errcode = vdRngGaussian(VSL_RNG_METHOD_GAUSSIAN_ICDF, get_rng_stream(), size, nvec, loc, d_one);
+            if (errcode != VSL_STATUS_OK)
+            {
+                throw std::runtime_error("DPNP RNG Error: dpnp_rng_noncentral_chisquare_c() failed.");
+            }
+            /* squaring could result in an overflow */
+            vmdSqr(size, nvec, nvec, VML_HA);
+            vmdAdd(size, result1, nvec, result1, VML_HA);
+
+            mkl_free(nvec);
+        }
+        else if (df < 1)
+        {
+            /* noncentral_chisquare(df, nonc) ~ G( df/2 + Poisson(nonc/2), 2) */
+            double lambda;
+            int* pvec = nullptr;
+            pvec = (int*)mkl_malloc(size * sizeof(int), 64);
+            if (pvec == nullptr)
+            {
+                throw std::runtime_error("DPNP RNG Error: dpnp_rng_noncentral_chisquare_c() failed.");
```

Comment:
DPNP alloc will take care about errors

Reply:
see, https://github.com/IntelPython/dpnp/pull/543#discussion_r571040391

Reply:
@shssf "DPNP alloc will take care about errors" not here but I have some checks for dpnp alloc on my code, I can generalize this work and make this changes on refactoring for all random functions. I have add this to TODO for refactoring.